velox/exec/tests/utils/Cursor.cpp

```diff
@@ -125,113 +125,242 @@ bool TaskQueue::hasNext() {
+    } else {
+      // NOTE: the destructor of 'executor_' will wait for all the async task
+      // activities to finish on TaskCursor destruction.
+      executor_ = std::make_shared<folly::CPUThreadPoolExecutor>(
+          std::thread::hardware_concurrency());
+      static std::atomic<uint64_t> cursorQueryId{0};
+      queryCtx_ = std::make_shared<core::QueryCtx>(
+          executor_.get(),
+          core::QueryConfig({}),
+          std::unordered_map<std::string, std::shared_ptr<Config>>{},
+          cache::AsyncDataCache::getInstance(),
+          nullptr,
+          nullptr,
+          fmt::format("SingleThreadedTaskCursorQuery_{}", cursorQueryId++));
+    }
+
+    if (!params.queryConfigs.empty()) {
+      auto configCopy = params.queryConfigs;
+      queryCtx_->testingOverrideConfigUnsafe(std::move(configCopy));
+    }
+
+    planFragment_ = {
+        params.planNode,
+        params.executionStrategy,
+        params.numSplitGroups,
+        params.groupedExecutionLeafNodeIds};
+    const std::string taskId =
+        fmt::format("test_single_threaded_cursor {}", ++serial_);
+
+    if (!params.spillDirectory.empty()) {
+      taskSpillDirectory_ = params.spillDirectory + "/" + taskId;
+      auto fileSystem =
+          velox::filesystems::getFileSystem(taskSpillDirectory_, nullptr);
+      VELOX_CHECK_NOT_NULL(fileSystem, "File System is null!");
+      try {
+        fileSystem->mkdir(taskSpillDirectory_);
+      } catch (...) {
+        LOG(ERROR) << "Faield to create task spill directory "
+                   << taskSpillDirectory_ << " base director "
+                   << params.spillDirectory << " exists["
+                   << std::filesystem::exists(taskSpillDirectory_) << "]";
+
+        std::rethrow_exception(std::current_exception());
+      }
+
+      LOG(INFO) << "Task spill directory[" << taskSpillDirectory_
+                << "] created";
+    }
   }
 
-  if (!params.queryConfigs.empty()) {
-    auto configCopy = params.queryConfigs;
-    queryCtx->testingOverrideConfigUnsafe(std::move(configCopy));
+ protected:
+  static std::atomic<int32_t> serial_;
+  std::shared_ptr<core::QueryCtx> queryCtx_;
+  core::PlanFragment planFragment_;
+  std::string taskSpillDirectory_;
+
+ private:
+  std::shared_ptr<folly::Executor> executor_;
+};
+
+std::atomic<int32_t> TaskCursorBase::serial_;
+
+class MultiThreadedTaskCursor : public TaskCursorBase {
+ public:
+  explicit MultiThreadedTaskCursor(const CursorParameters& params)
+      : TaskCursorBase(params),
+        maxDrivers_{params.maxDrivers},
+        numConcurrentSplitGroups_{params.numConcurrentSplitGroups},
+        numSplitGroups_{params.numSplitGroups} {
+    VELOX_DCHECK(!params.singleThreaded)
+
+    queue_ = std::make_shared<TaskQueue>(params.bufferedBytes);
+    // Captured as a shared_ptr by the consumer callback of task_.
+    auto queue = queue_;
+
+    const std::string taskId = fmt::format("test_cursor {}", ++serial_);
```

Comment:
Are we incrementing serial_ twice? here and in base class's constructor?

Reply:
In the [comment](https://github.com/facebookincubator/velox/pull/8406#discussion_r1457107991) it was suggested to use consistent task ID naming. Then I assume it's fine to share the same counter between 2 implementations?